### PR TITLE
feature: add market trade subscription in binance

### PR DESCRIPTION
**Pull request by zenixls2**
and update session.lastPrice accordingly.  

in backtest still doesn't work.  
also the field in market trade is very different from private trade. 
some fields in types.Trade is filled with dummy messages.

one missing part in types/exchange.go is added (the kucoin verification)

### CODE REVIEW

File: pkg/exchange/binance/parse.go [hunk 1]
```diff
@@ -462,6 +467,60 @@ func parseDepthEvent(val *fastjson.Value) (*DepthEvent, error) {
...
+type MarketTradeEvent struct {
+	EventBase
+	Symbol   string           `json:"s"`
+	Quantity fixedpoint.Value `json:"q"`
+	Price    fixedpoint.Value `json:"p"`
+
+	BuyerOrderId  int64 `json:"b"`
+	SellerOrderId int64 `json:"a"`
+
+	OrderTradeTime int64 `json:"T"`
```
Comment (c9s) [inline pkg/exchange/binance/parse.go:479]: I guess maybe you can use `types.MillisecondTimestamp` here? not quiet sure

File: pkg/exchange/binance/parse.go [hunk 2]
```diff
@@ -462,6 +467,60 @@ func parseDepthEvent(val *fastjson.Value) (*DepthEvent, error) {
...
+
+*/
+
+func (e *MarketTradeEvent) Trade() types.Trade {
+	tt := time.Unix(0, e.Time*int64(time.Millisecond))
+	return types.Trade{
+		ID:            uint64(e.TradeId),
+		Exchange:      types.ExchangeBinance,
+		Symbol:        e.Symbol,
+		OrderID:       uint64(e.BuyerOrderId),
```
Comment (c9s) [inline pkg/exchange/binance/parse.go:511]: use taker's order ID here?

File: pkg/exchange/binance/parse.go [hunk 3]
```diff
@@ -462,6 +467,60 @@ func parseDepthEvent(val *fastjson.Value) (*DepthEvent, error) {
...
+		Side:          types.SideTypeBoth,
```
Comment (c9s) [inline pkg/exchange/binance/parse.go:512]: Both is used for self trade type

Comment (zenixls2) [reply pkg/exchange/binance/parse.go:512]: shouldn't self trade be `SideTypeSelf`?  
for this case, we are not either buyer or seller, and we don't take our own order.
so should be `SideTypeBoth`.

File: pkg/exchange/binance/parse.go (hunk 1 above)
Comment (zenixls2) [reply pkg/exchange/binance/parse.go:479]: int64 is used elsewhere in the parse. Follow the same parsing interface of other events.

Comment (c9s) [reply pkg/exchange/binance/parse.go:479]: ok

File: pkg/exchange/binance/parse.go (hunk 2 above)
Comment (zenixls2) [reply pkg/exchange/binance/parse.go]: done

File: pkg/exchange/binance/parse.go [hunk 4]
```diff
@@ -462,6 +467,72 @@ func parseDepthEvent(val *fastjson.Value) (*DepthEvent, error) {
...
+	Quantity fixedpoint.Value `json:"q"`
+	Price    fixedpoint.Value `json:"p"`
+
+	BuyerOrderId  int64 `json:"b"`
+	SellerOrderId int64 `json:"a"`
+
+	OrderTradeTime int64 `json:"T"`
+	TradeId        int64 `json:"t"`
+
+	IsMaker bool `json:"m"`
```
Comment (c9s) [inline pkg/exchange/binance/parse.go:482]: you need to define the upper case `M` to fix the parsing.

Comment (zenixls2) [reply pkg/exchange/binance/parse.go:482]: done

**Review by c9s — APPROVED** [review]

**Pull request merged**